### DISCUSSION TITLE
Upload performance bottlenecked by server compared to Go for speed test (librespeed)?

### DISCUSSION THREAD

**Discussion by Zerowalker** [tokio-rs/axum#2823] [Q&A]
### Summary

For some reason uploading causes a bottleneck in the server (cpu usage) compared to go.
This is not me complaining, just wondering why as it seems a bit odd as I believe rust should have the advantage.
here is the go implementation: https://github.com/librespeed/speedtest-go

The upload part is (as far as i can tell) simply reading and discarding all data as fast as possible.

The relevant parts in go vs axum is this:

```go
func empty(w http.ResponseWriter, r *http.Request) {
	_, err := io.Copy(ioutil.Discard, r.Body)
	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
		return
	}
	_ = r.Body.Close()

	w.Header().Set("Connection", "keep-alive")
	w.WriteHeader(http.StatusOK)
}
```

This is my implementation of it, the headers are set using `tower_http::set_header::SetResponseHeaderLayer`,
but I have tried setting it directly in the function, results seem to be the same.

```rust
pub async fn empty(req: Request<Body>) -> StatusCode {
    let b: axum::body::BodyDataStream = req.into_body().into_data_stream();
    b.for_each(|_| async { () }).await;
    StatusCode::OK
}
```

The cpu usage on axum goes to 100% (on 1 thread) which limits the speed.
on go it's a lot less so it's not limited by it.

### axum version

0.7.5

**Comment by jplatte** [comment]
How are you running your Rust app?

**Reply by Zerowalker** [reply]
in release mode if that's what you meant?

**Reply by dayvejones** [reply]
Zerowalker, what is your OS? What is your hardware?

**Reply by Zerowalker** [reply]
Windows 10 x64 22H2 19045.4412
